odoo/osv/expression.py

```diff
@@ -331,6 +340,593 @@ def distribute_not(domain):
     return result
 
 
+def anyfy(domain, model):
+    """ Return the domain where all conditions on field sequences have been
+    transformed into 'any' conditions.
+    """
+    result = []
+    for item in domain:
+        if not is_operator(item) and not is_boolean(item):
+            left, operator, right = item
+            path = left.split('.', 1)
+            field = model._fields[path[0]]
+            if len(path) > 1 and field.relational:
+                subdomain = [(path[1], operator, right)]
+                comodel = model.env[field.comodel_name]
+                item = (path[0], 'any', anyfy(subdomain, comodel))
+            elif operator in ('any', 'not any'):
+                comodel = model.env[field.comodel_name]
+                item = (left, operator, anyfy(right, comodel))
+        result.append(item)
+    return result
+
+
+def make_tree(domain):
+    """ Return the domain as a tree, with the following structure::
+
+        <tree> ::= ('?', <boolean>)
+                |  ('!', <tree>) | ('&', [<tree]) | ('|', [<tree])
```

Comment:
Do you mean `('&', tree1, tree2, tree3)` instead of `('&', [tree1, tree2, tree3])` ?  That's a perfectly fine alternative.
I find it slightly easier to parse this way, but it's mostly a matter of taste.

Reply:
We opted for n-ary sequence, what do you think @xmo-odoo ?